### PR TITLE
Laurentvallet issue#1980 (qc in ShUp)

### PR DESCRIPTION
**Pull request by LaurentPV**
Allowing ShanoirUploader to execute quality-control check using datasets ms code before executing the import.

Check quality code externalized in a new QualityService class in datasets.

Same rules as quality check on server side.

Apply only if quality card is checkAtImport.

Returns qualityCardResult report if in error, warning or failed valid.

Update upload status in ERROR if quality check in error.

### CODE REVIEW

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/importer/service/ImporterService.java [hunk 1]
```diff
@@ -167,18 +174,37 @@ public void createAllDatasetAcquisition(ImportJob importJob, Long userId) throws
...
+                            subjectStudy.setQualityTag(qualityResult.get(0).getTagSet());
+                            qualityResult.addUpdatedSubjectStudy(subjectStudy);
+                        }
+                    }
+                }
+                                				
                 // Has quality check passed ?
-                if (qualityResult.hasError()) {
+                if (qualityResult != null && !qualityResult.isEmpty() && qualityResult.hasError()) {
+                    // TODO : Delete newly created Examination ?
```
Comment (julien-louis) [inline shanoir-ng-datasets/src/main/java/org/shanoir/ng/importer/service/ImporterService.java:197]: I think deleting the exam would not fit to the frontend import workflow as the exam is created "independently"

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/model/condition/StudyCardDICOMConditionOnDatasets.java [hunk 2]
```diff
@@ -229,16 +229,6 @@ private Boolean fulfilled(Attributes dicomAttributes, StringBuffer errorMsg, Obj
                         return true; // as condition values are combined by OR: return if one is true
                     } // else continue to check other values
     
-                } else if (DicomTagType.FloatArray.equals(tagType)) {
```
Comment (julien-louis) [inline shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/model/condition/StudyCardDICOMConditionOnDatasets.java:232]: :+1:

**Review by julien-louis — COMMENTED** [review]
I have no problems with the code, now I must do some tests

**Comment by julien-louis** [comment]
@LaurentPV  as said in the chat, failing to check quality should not bypass it. It should result in stopping the import and prompting an explicit error message.

**Comment by julien-louis** [comment]
Hi @LaurentPV  I'm vey sorry I take so much time to review this but could you lease merge develop to your branch there are 2 conflicts that I don't know how to fix.

**Review by julien-louis — CHANGES_REQUESTED** [review]
I have a bug when a study has no quality card, it blocks the import.

```
2024-11-12 11:35:17 ERROR [org.shanoir.uploader.action.ImportDialogOpener] [77] Cannot invoke "java.util.Collection.toArray()" because "c" is null
java.lang.NullPointerException: Cannot invoke "java.util.Collection.toArray()" because "c" is null
	at java.base/java.util.ArrayList.addAll(ArrayList.java:752)
	at org.shanoir.uploader.action.ImportDialogOpener.getAllQualityCards(ImportDialogOpener.java:349)
	at org.shanoir.uploader.action.ImportDialogOpener.getStudiesWithStudyCards(ImportDialogOpener.java:119)
	at org.shanoir.uploader.action.ImportDialogOpener.openImportDialog(ImportDialogOpener.java:60)
	at org.shanoir.uploader.nominativeData.CurrentNominativeDataController$1.mouseClicked(CurrentNominativeDataController.java:108)
	at java.desktop/java.awt.AWTEventMulticaster.mouseClicked(AWTEventMulticaster.java:278)
``` 
```
private List<QualityCard> getAllQualityCards(List<Study> studies) throws Exception {
		List<QualityCard> qualityCards = new ArrayList<QualityCard>();
		for (Iterator<Study> iterator = studies.iterator(); iterator.hasNext();) {
			Study study = (Study) iterator.next();
				qualityCards.addAll(shanoirUploaderServiceClient.findQualityCardsByStudyId(study.getId()));
		}
		return qualityCards;
	}
```

**Comment by julien-louis** [comment]
Also I wonder what is the meaning of this field that shows only auto check quality cards : 
image

Does this mean you can't have several quality cards checks on imports ? I think the back checks every quality card that is set as auto-check.

**Comment by julien-louis** [comment]
3. Sorry I can't make the uploader to block an import. The import goes well with my quality card here : 
image

**Comment by michaelkain** [comment]
to see by Laurent

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/importer/service/ImporterService.java (hunk 1 above)
Comment (LaurentPV) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/importer/service/ImporterService.java:197]: I was worried about the empty created exams if quality control is KO and the fact that the user may recreated a new exam each time he tries to reimport the data.

File: shanoir-uploader/pom.xml [hunk 3]
```diff
@@ -39,6 +40,17 @@
 			    </exclusion>
 			</exclusions>
 		</dependency>
+		<dependency>
+			<groupId>org.shanoir-ng</groupId>
+			<artifactId>shanoir-ng-datasets</artifactId>
+			<version>3.2.2</version>
+			<exclusions>
+			    <exclusion>
+			        <groupId>*</groupId>
```
[the remark this answers is not in the document]
Comment (LaurentPV) [reply shanoir-uploader/pom.xml:49]: It excludes all dependencies from the dependency ms-datasets to avoid dependency loops

**Comment by LaurentPV** [comment]
Bugfix if no quality cards found is pushed. Branch is up to date.

**Review by julien-louis — APPROVED** [review]

**Pull request merged**